idx/memory/tag_query.go

```diff
@@ -171,247 +239,812 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
 			// must be present in the byId lookup table
 			corruptIndex.Inc()
 			log.Error(3, "memory-idx: ID %q is in tag index but not in the byId lookup table", id.String())
-			delete(resultSet, id)
 			continue
 		}
 
-		if def.LastUpdate < q.from {
-			delete(resultSet, id)
+		if q.testByAllExpressions(id, def, false) {
+			resCh <- id
 		}
 	}
-}
 
-func (q *TagQuery) Run(index TagIndex, byId map[string]*idx.Archive) TagIDs {
-	var resultSet TagIDs
+	q.wg.Done()
+	completeCh <- struct{}{}
+}
 
+// sortByCost tries to estimate the cost of different expressions and sort them
+// in increasing order
+// this is to reduce the result set cheaply and only apply expensive tests to an
+// already reduced set of results
+func (q *TagQuery) sortByCost() {
 	for i := range q.equal {
-		q.equal[i].cost = uint(len(index[q.equal[i].key][q.equal[i].value]))
+		q.equal[i].cost = uint(len(q.index[q.equal[i].key][q.equal[i].value]))
+	}
+
+	for i := range q.prefix {
+		q.prefix[i].cost = uint(len(q.index[q.prefix[i].key][q.prefix[i].value]))
 	}
 
 	for i := range q.match {
-		q.match[i].cost = uint(len(index[q.match[i].key]))
+		q.match[i].cost = uint(len(q.index[q.match[i].key]))
 	}
 
 	sort.Sort(KvByCost(q.equal))
 	sort.Sort(KvByCost(q.notEqual))
+	sort.Sort(KvByCost(q.prefix))
 	sort.Sort(KvReByCost(q.match))
 	sort.Sort(KvReByCost(q.notMatch))
+}
 
-	if q.startWith == EQUAL {
-		resultSet = q.getInitialByEqual(index, q.equal[0])
-		q.equal = q.equal[1:]
+// Run executes the tag query on the given index and returns a list of ids
+func (q *TagQuery) Run(index TagIndex, byId map[string]*idx.Archive) TagIDs {
+	q.index = index
+	q.byId = byId
+
+	q.sortByCost()
+
+	idCh, _ := q.getInitialIds()
+	resCh := make(chan idx.MetricID)
+	completeCh := make(chan struct{}, tagQueryWorkers)
+
+	// start the tag query workers. they'll consume the ids on the idCh and
+	// evaluate for each of them whether it satisfies all the conditions
+	// defined in the query expressions. those that satisfy all conditions
+	// will be pushed into the resCh
+	// when a worker completes it pushes an empty item into the completeCh to
+	// signal its completion
+	q.wg.Add(tagQueryWorkers)
+	for i := 0; i < tagQueryWorkers; i++ {
+		go q.filterIdsFromChan(idCh, resCh, completeCh)
+	}
+
+	completedWorkers := 0
+	result := make(TagIDs)
+
+IDS:
+	for {
+		select {
+		case id := <-resCh:
+			result[id] = struct{}{}
+		case <-completeCh:
+			completedWorkers++
+			if completedWorkers >= tagQueryWorkers {
+				// once all workers have completed the query execution is done.
+				// we need to consume the result channel until it's empty before
+				// returning the results
+				for {
+					select {
+					case id := <-resCh:
+						result[id] = struct{}{}
+					default:
+						break IDS
+					}
+				}
+			}
+		}
+	}
+
+	q.wg.Wait()
+	return result
+}
+
+// getMaxTagCount calculates the maximum number of results a tag query could
+// possibly return
```

Comment:
this is called cardinality

Reply:
really? i thought cardinality is how many unique values each tag has. but this function returns the max number of tags

Reply:
you can have cardinality of data and cardinality of a resultset, both simply convey the number of distinct values

Reply:
0f0bbc1